refactor:  Connectivity Documentation (#1616)
SAP/cloud-sdk: docs-java/features/connectivity/004-http-destinations.mdx

**However**, there might be properties that contain non-primitive values, such as lists, maps, or other objects.
These values **might** be mutable, depending on their implementation.
Therefore, it is impossible to guarantee _true immutability_ for the `DefaultHttpDestination` implementation.
Instead, the implementation provides a _shallow immutability_ guarantee.

This implementation detail is important for using [the builder API](#about-the-builder-api).

## About HTTP Headers

One of the most important properties of any `HttpDestination` are the HTTP headers, which are attached to any outgoing request performed with the destination.

The `DefaultHttpDestination` implements support for following sources of HTTP headers:

1. Properties named using the pattern `URL.header.<header-name>` will be converted into HTTP headers.
2. [Special properties](#special-header-providers), which will be converted into HTTP headers.
3. Headers that were [directly added](#adding-individual-headers) via the [Builder API](#about-the-builder-api) while assembling the `DefaultHttpDestination` instance.
4. Headers that are provided by [`DestinationHeaderProvider`s](#about-headerproviders).

> 5. Any headers required for on-premise connectivity will automatically be added, **iff** the `ProxyType` property is set to `OnPremise` as [explained below](#on-premise-handling).

Reviewer: ```suggestion
5. Any headers required for on-premise connectivity will automatically be added, **iff** the `ProxyType` property is set to `OnPremise` as [explained below](#deriving-the-proxy-authorization-header).
```

### Deriving The `Authorization` Header

The `DefaultHttpDestination` tries to derive the `Authorization` header from authentication related properties, **iff** no `Authorization` header is provided by any other source.
In most cases, however, the BTP Destination Service will [provide a `Authorization` header](btp-destination-service#how-authentication-at-the-target-system-works), so this step is usually skipped.

If the destination does not stem from the BTP destination service, then the concrete properties depend on the `AuthenticationType` of the `DefaultHttpDestination`.

- For `AuthenticationType.BASIC_AUTHENTICATION`, the basic credentials are taken from the `User` (or `username`, if the former doesn't exist) and the `Password` properties. <br />
  If these properties are not set, an exception is thrown.

- For `AuthenticationType.TOKEN_FORWARDING`, the `RequestHeaderAccessor` is used to find any `Authorization` header in the current request. All headers that are found (there should usually only be one) are then forwarded.

### Deriving The `Proxy-Authorization` Header

:::info On-Premise Handling

> As described [above](#on-premise-handling), the `DefaultHttpDestination` implements dedicated support for on-premise connectivity.

Reviewer: ```suggestion
As described above, the `DefaultHttpDestination` implements dedicated support for on-premise connectivity.
```

Current link is dead.

2. [Special properties](#special-header-providers), which will be converted into HTTP headers.
3. Headers that were [directly added](#adding-individual-headers) via the [Builder API](#about-the-builder-api) while assembling the `DefaultHttpDestination` instance.
4. Headers that are provided by [`DestinationHeaderProvider`s](#about-headerproviders).
5. (Conditional) A `Proxy-Authorization` (and potentially a `SAP-Connectivity-SCC-Location_ID`) HTTP header will automatically be added, **iff** the `ProxyType` property is set to `OnPremise` as [explained below](#on-premise-handling).
6. (Conditional) Headers that are dervied from [authentication related properties](#deriving-the-authorization-header), **iff** no `Authorization` header was provided by any of the previous options.
7. (Conditional) Headers that are derived from [proxy authentication related properties](#deriving-the-proxy-authorization-header), **iff** no `Proxy-Authorization` header was provided by any of the previous options.

### On-Premise Handling

The `DefaultHttpDestination` implements dedicated support for on-premise connectivity as [documented here](on-premise).

### Deriving The `Authorization` Header

> As [explained above](#about-http-headers), the `DefaultHttpDestination` tries to derive the `Authorization` header from authentication related properties, **iff** no `Authorization` header is provided by any other source.

Reviewer 2: ```suggestion
The `DefaultHttpDestination` tries to derive the `Authorization` header from authentication related properties, **iff** no `Authorization` header is provided by any other source.
```

(minor) I feel like there is not much value in referencing text that is 3-4 sentences away above 😉

Reviewer 2: (minor) Also, this would be a good place to mention those typically come from the BTP destination service, linking to that doc?

## About The _Builder_ API

:::danger Expert API

Using the _Builder_ API requires a deep understanding of the `DefaultHttpDestination` implementation.
Therefore, it is considered an _expert API_ and should only be used if absolutely necessary. <br />
We encourage users to rather use [the `DestinationAccessor` API](btp-destination-service) or [the `ServiceBindingDestinationLoader` API](service-bindings) to retrieve destinations.

:::

The [`DefaultHttpDestination.Builder`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html) allows you to create a `DefaultHttpDestination` instance from scratch.
It allows you to set any desired property, header, header provider, or any other value that is supported by the `DefaultHttpDestination` implementation.

### Building On Top Of An Existing Destination

Sometimes, you may want to slightly adjust an existing `Destination` instance.
For example, there might be just this one single header missing for a specific request. <br />

> Unfortunately, adding this header directly to the existing `Destination` instance is not possible, as [the `Destination` interface is immutable](#about-immutability).

Reviewer 2: (minor) Personally, I would omit terms like unfortunately / luckily to keep the type of language consistent

</details>

### Adding Individual Headers

Individual headers can be added via the [`header(Header)`](<pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html#header(com.sap.cloud.sdk.cloudplatform.connectivity.Header)>) and [`header(String, String)`](<pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html#header(java.lang.String,java.lang.String)>) methods.
In either case, the newly added header is **only added** to the list of all headers.
It will **not** override any existing header with the same name.

### Adding Individual Header Providers

The [`headerProviders(DestinationHeaderProvider...)`](<pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html#headerProviders(com.sap.cloud.sdk.cloudplatform.connectivity.DestinationHeaderProvider...)>) method allows you to add as many `DestinationHeaderProvider`s to the `DefaultHttpDestination` instance as you wish.
Once again, existing header providers will **not** be overridden.

### Adding Static Header Providers

> In some scenarios, there are certain HTTP headers that should always be sent to a destination, regardless of the request.

Reviewer 2: (minor) Technically this is unrelated to the Builder API. Since we have a dedicated section about header providers below, maybe just move it there?
